certstore_linux.go

```diff
@@ -1,14 +1,318 @@
+		key  = C.PK11_FindKeyByAnyCert(&cert, nil)
+	)
+	if key == nil {
+		return nil, errors.New("cannot find private key")
+	}
+
+	digestC := C.SECITEM_AllocItem(nil, nil, C.uint(len(digest)))
+	if digestC == nil {
+		return nil, errors.New("failure to allocate memory for digest")
+	}
+	defer C.SECITEM_FreeItem(digestC, 1)
+
+	resultC := C.SECITEM_AllocItem(nil, nil, C.uint(C.PK11_SignatureLen(key)))
+	if resultC == nil {
+		return nil, errors.New("failure to allocate memory for signature")
+	}
+	defer C.SECITEM_FreeItem(resultC, 1)
+
+	digestC.len = C.uint(len(digest))
+	C.memcpy(unsafe.Pointer(digestC.data), unsafe.Pointer(&digest[0]), C.size_t(len(digest)))
+
+	if C.SGN_Digest(key, goToNssAlg[hash], resultC, digestC) != C.SECSuccess {
+		if C.PORT_GetError() == C.SEC_ERROR_INVALID_ALGORITHM {
+			return nil, ErrUnsupportedHash
+		}
+		return nil, fmt.Errorf("error signing: %s\n", C.GoString(C.GetErrorString()))
+	}
+
+	if key.keyType == C.ecKey {
+		var dsaSig C.SECItem
+		if C.DSAU_EncodeDerSigWithLen(&dsaSig, resultC, resultC.len) != C.SECSuccess {
+			return nil, errors.New("unable to generate dsa signature")
+		}
+		resultC = &dsaSig
+	}
+	return C.GoBytes(unsafe.Pointer(resultC.data), C.int(resultC.len)), nil
+}
+
+// Import implements the Store interface.
+func (nssStore) Import(data []byte, password string) error {
+	unicodePassword, err := bmpString(password)
+	if unicodePassword == nil {
+		return err
+	}
+	pass := C.SECITEM_AllocItem(nil, nil, C.uint(len(unicodePassword)))
+	if pass == nil {
+		return errors.New("SECITEM_AllocItem failed")
+	}
+	defer C.SECITEM_FreeItem(pass, 1)
+	C.memcpy(unsafe.Pointer(pass.data), unsafe.Pointer(&unicodePassword[0]), C.size_t(len(unicodePassword)))
+	var (
+		p12     = C.SEC_PKCS12DecoderStart(pass, nil, nil, nil, nil, nil, nil, nil)
+		decoded = C.SEC_PKCS12DecoderUpdate(p12, (*C.uchar)(unsafe.Pointer(&data[0])), C.size_t(len(data)))
+	)
+	if decoded != 0 {
+		return fmt.Errorf("error %d, P12 decoding failed", int(C.PR_GetError()))
+	}
+	authenticated := C.SEC_PKCS12DecoderVerify(p12)
+	if authenticated != 0 {
+		return fmt.Errorf("error %d, P12 authentication failed", int(C.PR_GetError()))
+	}
+	validated := C.SEC_PKCS12DecoderValidateBags(p12, (*[0]byte)(C.P12U_NicknameCollisionCallback))
+	if validated != 0 {
+		return fmt.Errorf("error %d, P12 validation failed", int(C.PR_GetError()))
+	}
+	imported := C.SEC_PKCS12DecoderImportBags(p12)
+	if imported != 0 {
+		return fmt.Errorf("error %d, P12 import failed", int(C.PR_GetError()))
+	}
+	return nil
+}
+
+// Close implements the Store interface.
+func (ctx nssStore) Close() {
+	nssCtx := C.NSSInitContext(ctx)
+	C.NSS_ShutdownContext(&nssCtx)
+}
+
+// openStore opens the current user's NSS database.
 func openStore() (Store, error) {
-	return nil, errors.New("certstore only works on macOS and Windows")
+	homeDir, err := os.UserHomeDir()
+
+	if err != nil {
+		return nil, err
+	}
+
+	nssdb := path.Join(homeDir, ".pki", "nssdb")
```

Comment:
All the docs I'm seeing on macos say that the default directory is `~/.netscape`. It would be good to make this at least somewhat configurable. I'll open a separate PR to allow `openStore` to receive configuration arguments.

Reply:
I'm not sure if it's this simple unfortunately - I don't have a `~/.netscape`, for example

Reply:
Gotcha. I had been thinking it would be normal for multiple applications to share this data, but I guess that isn't the case.

Reply:
I think we can get away with being a "Mode 3, Type B" application as described here: https://wiki.mozilla.org/NSS_Shared_DB

Unfortunately, all the examples there are for other scenarios :face_with_head_bandage:

Reply:
Also note that NSS databases can be password protected, so that will be fun to deal with

Reply:
As far as I can tell, in general nobody attempts to have a system-wide NSS database and instead apps have a local one. Whilst it's possible for apps to share an NSS database in the new SQLite world, I still don't see apps actually doing it (there's a [HOWTO guide](https://wiki.mozilla.org/NSS_Shared_DB_Howto) explaining how to set it up manually, involving making one in ~/.pki/nssdb and then symlinking it into the firefox & thunderbird profile directories). So, I think probably apps using certstore e.g. smimesign should take care of deciding where the database should live.

It's also not clear to me whether uses of hardware PKCS#11 tokens even need a database - the [FAQ](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/PKCS11/FAQ) suggests that everything happens in an in-memory database (see "Will NSS get the user's CA certificate via PKCS #11 and push it into the CA certificate database or is the CA certificate database expected to obtain the CA certificate by some other means?") so possibly we only need to use an on-disk DB if the soft ware PKCS#11 library (I think it's called softokn) is being used.

Reply:
Well, that's good news regarding PKCS#11 tokens at least. Applications having to maintain their own database isn't ideal, but isn't the end of the world either.